flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/MultipleInputStreamTask.java

```diff
@@ -244,6 +235,29 @@ protected void createInputProcessor(
         return resultFuture;
     }
 
+    private CompletableFuture<Boolean> triggerStopWithSavepointWithDrainAsync(
+            CheckpointMetaData checkpointMetaData, CheckpointOptions checkpointOptions) {
+        mainMailboxExecutor.execute(
+                () -> {
+                    setSynchronousSavepoint(checkpointMetaData.getCheckpointId(), true);
+                    operatorChain.getSourceTaskInputs().forEach(s -> s.getOperator().stop());
+                },
+                "stop chained Flip-27 source for stop-with-savepoint --drain");
+
+        CompletableFuture<Void> sourcesStopped =
+                FutureUtils.waitForAll(
+                        operatorChain.getSourceTaskInputs().stream()
+                                .map(s -> s.getOperator().getFinishedFuture())
+                                .collect(Collectors.toList()));
```

Comment:
Maybe it would be better to collect the futures inside the mailbox action `mainMailboxExecutor.submit()`? And keep the previous signature of `SourceOperator#stop()` returning `finished` future? 

As it is we are here assuming `operatorChain.getSourceTaskInputs()` and `s.getOperator().getFinishedFuture()` are thread safe, which I think they are (because they are immutable), but it's not part of the guarantees.